src/controllers/water-level-changes.js

```diff
@@ -0,0 +1,132 @@
+    recentWaterLevel = updatedData[0].water_level_m
+  } catch (e) {
+    e.code = 500
+    throw e
+  }
+
+  let responseValues
+  if (storedUpdated && new Date(storedUpdated) >= new Date(recentUpdated)) {
+    console.log('No new updates')
+    responseValues = await req.cache.getAll(KEY, req.body['limit'])
+  } else {
+    console.log('Dataset was updated')
+    let oneDayBounds = getBounds(recentUpdated, 1)
+    let threeDaysBounds = getBounds(recentUpdated, 3)
+    let oneWeekBounds = getBounds(recentUpdated, 7)
+    let oneDayResults
+    let threeDaysResults
+    let oneWeekResults
+    try {
+      await Promise.all(
+        (oneDayResults = await getData(oneDayBounds)),
+        (threeDaysResults = await getData(threeDaysBounds)),
+        (oneWeekResults = await getData(oneWeekBounds))
+      )
+    } catch (e) {
+      e.code = 500
+      throw e
+    }
+    console.log('Adding new rows')
+    let id = uuid()
+    let newRows = {
+      id,
+      created_at: new Date(Date.now()).toISOString(),
+      last_updated: recentUpdated,
+      latest_water_level: recentWaterLevel,
+      one_day_average: oneDayResults[0].toFixed(3),
+      one_day_delta: oneDayResults[1].toFixed(3),
+      three_day_average: threeDaysResults[0].toFixed(3),
+      three_day_delta: threeDaysResults[1].toFixed(3),
+      one_week_average: oneWeekResults[0].toFixed(3),
+      one_week_delta: oneWeekResults[1].toFixed(3),
+      meta: {
+        id,
+        timestamp: Math.round(new Date() / 1000)
+      }
+    }
+    req.cache.add(KEY, newRows)
+    responseValues = await req.cache.getAll(KEY, req.body['limit'])
+  }
+
+  res.status(200).send({
+    data: responseValues
+  })
+}
+
+/**
+ * This function obtains the upper and lower bounds of a time period
+ * @param {String} date The date in ISO format
+ * @param {Number} days The number of days the period should be
+ * @returns {Array<String>} The upper and lower bounds of the time period
+ */
+function getBounds(date, days) {
+  const dateRegex = /[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}.[0-9]{3}/
+  let currentDate = new Date(new Date(date) - TIMEZONE_OFFSET_MILLIS)
+  let threeDayPeriod = new Date(
+    new Date(date) - (days * 24 * 60 * 60 * 1000 + TIMEZONE_OFFSET_MILLIS)
+  )
+  let lowerBound = threeDayPeriod.toISOString().match(dateRegex)[0]
+  let upperBound = currentDate.toISOString().match(dateRegex)[0]
+  return [lowerBound, upperBound]
+}
+
+/**
+ * Returns the average and delta of the water levels respecitvely in an array
+ * @param {Array<String>} bounds The bounds to do calculation on (dates in ISO 8601 format)
+ * @returns {Array<Number>} The average and delta of the water level between the specified bounds
+ */
+async function getData(bounds) {
+  let workingDataQuery = `${QUERY_BASE}
+    SELECT water_level_m
+    WHERE station_number = '05DF001'
+    AND date_and_time
+    BETWEEN '${bounds[0]}'
+    AND '${bounds[1]}'
+    ORDER BY date_and_time DESC
+    LIMIT 10000`
```

Comment:
We may need to think about this. Returning 10000 rows is going to be slow. Can we cache this information somehow?

Reply:
Caching the already obtained data sounds like a good idea

Reply:
Caching is very possible. We could write a storage interface, call it `rolling-average.js` or something. We never request the same results twice. When a new result comes in, we can update our averages using [this](https://math.stackexchange.com/questions/106700/incremental-averageing) approach.

Reply:
I moved the calculation for the average water levels of each time period to the SoQL queries. The query should now only return 1 row: the number of data points involved in the average calculation and the average water level itself.

Reply:
The query referenced can be found [here](https://github.com/WJXHenry/IFTTT-Edmonton/blob/09184a30e7bc5c1908020dd35ff3d61e9fa945e3/src/controllers/water-level-changes.js#L169-L188).